cpp/unreal_plugins/SimulationController/Source/SimulationController/UrdfBotAgent.cpp

```diff
@@ -0,0 +1,157 @@
+    actor_spawn_params.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+    urdf_bot_pawn_ = world->SpawnActor<AUrdfBotPawn>(FVector::ZeroVector, FRotator::ZeroRotator, actor_spawn_params);
+    ASSERT(urdf_bot_pawn_);
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "camera")) {
+        camera_sensor_ = std::make_unique<CameraSensor>(
+            urdf_bot_pawn_->camera_component_,
+            Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.RENDER_PASSES"),
+            Config::get<unsigned int>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.IMAGE_WIDTH"),
+            Config::get<unsigned int>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.IMAGE_HEIGHT"));
+        ASSERT(camera_sensor_);
+
+        // update FOV
+        for (auto& pass : Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.RENDER_PASSES")) {
+            camera_sensor_->render_passes_.at(pass).scene_capture_component_->FOVAngle =
+                Config::get<float>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.FOV");
+        }
+    }
+}
+
+UrdfBotAgent::~UrdfBotAgent()
+{
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "camera")) {
+        ASSERT(camera_sensor_);
+        camera_sensor_ = nullptr;
+    }
+
+    ASSERT(urdf_bot_pawn_);
+    urdf_bot_pawn_->Destroy();
+    urdf_bot_pawn_ = nullptr;
+}
+
+void UrdfBotAgent::findObjectReferences(UWorld* world) {}
+
+void UrdfBotAgent::cleanUpObjectReferences() {}
+
+std::map<std::string, Box> UrdfBotAgent::getActionSpace() const
+{
+    std::map<std::string, Box> action_space;
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "joint")) {
+        for (auto& joint_component : urdf_bot_pawn_->urdf_robot_component_->joint_components_) {
+            if (joint_component.second->control_type_ != UrdfJointControlType::Invalid) {
+                Box box;
+                box.low_ = std::numeric_limits<float>::lowest();
+                box.high_ = std::numeric_limits<float>::max();
+                box.shape_ = {1};
+                box.datatype_ = DataType::Float32;
+                action_space[joint_component.first] = std::move(box);
+            }
+        }
+    }
+
+    return action_space;
+}
+
+std::map<std::string, Box> UrdfBotAgent::getObservationSpace() const
+{
+    std::map<std::string, Box> observation_space;
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "link_state")) {
```

Comment:
Since we've already gone half-way to implementing this in the cleanest possible way, we might as well go all the way. Let's mimic how `CameraSensor` does this, i.e., let's have a `getObservationSpace(observation_components)` method in `UrdfRobotComponent`. We'll need to move `Rpclib.h` and `Box.h` into `CoreUtils`.

Reply:
fixed